### PR TITLE
BUG: optimizing compilers can reorder call to npy_get_floatstatus

### LINKED ISSUE #10370: BUG: np.min does not always propagate NaNs
This investigation arises from scipy CI starting to [fail on appveyor](https://ci.appveyor.com/project/scipy/scipy/build/1.0.1444/job/n05ptntm0xxjklvt) in the last few days, i.e. after 1.14 being released.

On my home computer (macOS, conda python 3.6.2, conda numpy):
```
>>> import numpy as np
>>> np.version.version
'1.14.0'
>>> np.min([1.0, 2.0, np.nan])
nan
>>> np.min([1.0, np.nan, 2.0])
nan
>>> np.min([np.nan, 1.0, 2.0])
nan
>>> np.min([np.nan, 1.0])
/Users/andrew/miniconda3/envs/dev3/lib/python3.6/site-packages/numpy/core/_methods.py:29: RuntimeWarning: invalid value encountered in reduce
  return umr_minimum(a, axis, None, out, keepdims)
nan
```

On my work computer (macOS, conda python 3.6.2, numpy installed via pip in a clean env):
```
>>> import numpy as np
>>> np.version.version
'1.14.0'
>>> np.min([1., 2., 3., 4., np.nan])
nan
>>> np.min([1., 2., 3., np.nan, 4.])
nan
>>> np.min([1., 2., np.nan, 3., 4.])
nan
>>> np.min([1., np.nan, 2., 3., 4.])
nan
>>> np.min([np.nan, 1., 2., 3., 4.])
nan
>>> np.min([np.nan, 1.])
nan
>>> np.min([np.nan, 1., np.nan])
nan
>>> np.min([1., np.nan])
nan
>>> np.seterr(all='raise')
{'divide': 'warn', 'over': 'warn', 'under': 'ignore', 'invalid': 'warn'}
>>> np.min([1., np.nan])
nan
>>> np.min([np.nan, 1.])
nan
>>> np.min([np.nan, 1., 2., 3., 4.])
nan
>>> np.min([np.nan, 1., 2., 3., 4.])
nan
```

With the first set of code examples why don't the first three examples result in a warning, only the last?
With the second set of examples no warning is emitted at all.

### PR DESCRIPTION
**Pull request by mattip**
Fixes #10370. We should find a more generic and explicit way to prevent optimizing compilers from reordering the call to `npy_get_floatstatus`.

To reproduce the problem, clang or gcc-8.1 are required. Confirmed that this fixes the problem using clang-6.0

### CODE REVIEW

**Comment by eric-wieser** [comment]
Does putting the `volatile` in `npy_get_floatstatus` instead work as well?

**Comment by eric-wieser** [comment]
I don't understand that test - what is `outside`?

**Comment by mattip** [comment]
`outside` is a dummy function, the equivalent of `npy_get_floatstatus`. It's implementation is irrelevant to the compilation of this unit

**Comment by eric-wieser** [comment]
>  Its implementation is irrelevant to the compilation of this unit

But it's not - in the numpy case, the compiler is (probably?) able to inline the function, whereas here it isn't able to.

**Comment by tzickel** [comment]
A. I don't know if people usually build numpy wth LTO, but it's worth checking that if using LTO, the compiler might see the internals of that function and decide that it's the safe to still reorder it.

B. The rest of the npy_*floatstatus* usage in the code needs to be audited on clang / gcc and see that the compiler does not reorder it in a bad way as well.

C. Testing it on MSVC seeing it's still ok (there in the documentation it says that using FENV_ACCESS is the way to go, maybe detect compiler and do something differently).

**Comment by eric-wieser** [comment]
> If I add c1 as a parameter to outside, the compiler generates the correct assembler 

That's not what I see when I click that link:

image

The first thing that happens is your `outside` is called.

**Comment by eric-wieser** [comment]
What is working for me is adding a volatile `char` in `npy_get_floatstatus`, which then forces the argument to be passed:

https://godbolt.org/g/gEqR6P

**Comment by juliantaylor** [comment]
`npy_get_float_status` is not const function, the compiler is not allowed to reorder it as it can (and does) depends on the global state.
This looks like a compiler bug we should investigate. Or clang has changed some options so the floating point state is not considered anymore by default.

Is it reproducable with gcc 8? I tried with 8.0.1 and the code looks correct.

**Comment by mattip** [comment]
@juliantaylor according to https://godbolt.org/g/gEqR6P the issue exists in old versions of clang, and now reproduces in gcc 8.1

**Comment by mattip** [comment]
is my fix considered a C-API change? Neither scipy nor cython use the `npy_.*floatstatus.*`functions

File: numpy/core/src/npymath/ieee754.c.src [hunk 1]
```diff
@@ -759,13 +780,18 @@ int npy_clear_floatstatus(void)
 
 #else
 
-int npy_get_floatstatus(void)
+int npy_get_floatstatus(void *param)
```
Comment (charris) [inline numpy/core/src/npymath/ieee754.c.src]: Why pass as `void *` instead of `char *`?

**Comment by charris** [comment]
> is my fix considered a C-API change?

Yes, and I think you should not use the same function names if you change the signatures because that will break current code, not just require a recompile. You could make it feature version dependent, but then folks would be stuck with using at least that version of numpy. See `npy_config.h` and `ndarraytypes.h`. AFAICT, we have never used the `NPY_FEATURE_VERSION` by itself. I suppose one way to avoid problems would be to make the new functions private, but they may be needed downstream. 

Any particular reason to pass the arguments as `void *` ?

**Comment by juliantaylor** [comment]
GCC bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85633

File: numpy/core/src/npymath/ieee754.c.src (hunk 1 above)
Comment (mattip) [reply numpy/core/src/npymath/ieee754.c.src]: no particular reason. This way any pointer argument will suffice and silently cast, if it is `char*` I will need to cast everywhere it is called, no?

**Comment by mattip** [comment]
So if this approach is acceptable, I will change the function names.

File: doc/source/reference/c-api.coremath.rst [hunk 2]
```diff
@@ -150,31 +150,31 @@ Those can be useful for precise floating point comparison.
 
     .. versionadded:: 1.4.0
 
-.. c:function:: void npy_set_floatstatus_divbyzero()
+.. c:function:: void npy_set_floatstatus_divbyzero(void*)
```
Comment (juliantaylor) [inline doc/source/reference/c-api.coremath.rst]: do we need barriers on the set functions?
reordering becomes interesting when reading the status but setting doesn't really matter as it has no influence on future instructions.

**Comment by ahaldane** [comment]
I'm not sure yet about 1.14.4, but let's tag things to see what would go in it.

**Comment by tzickel** [comment]
nice throw back from 7 years ago:
https://github.com/numpy/numpy/blame/b946795bd64d34147f3486e97246ed1f6a4a0937/numpy/core/src/umath/scalarmath.c.src#L809

:)

**Comment by juliantaylor** [comment]
the scalarmath code could also be updated to use the functions with the barriers.

**Comment by charris** [comment]
Segmentation fault.

**Comment by charris** [comment]
```
  numpy\core\src\npymath\ieee754.c.src(754) : error C2143: syntax error : missing ';' before 'type'
  numpy\core\src\npymath\ieee754.c.src(761) : error C2065: 'fpstatus' : undeclared identifier
  numpy\core\src\npymath\ieee754.c.src(762) : error C2065: 'fpstatus' : undeclared identifier
  numpy\core\src\npymath\ieee754.c.src(763) : error C2065: 'fpstatus' : undeclared identifier
  numpy\core\src\npymath\ieee754.c.src(764) : error C2065: 'fpstatus' : undeclared identifier
```
Also unused variable warnings.

File: doc/source/reference/c-api.coremath.rst [hunk 3]
```diff
@@ -185,12 +185,33 @@ Those can be useful for precise floating point comparison.
...
+
+    .. versionadded:: 1.15.0
+
 .. c:function:: int npy_clear_floatstatus()
 
     Clears the floating point status. Returns the previous status mask.
 
     .. versionadded:: 1.9.0
 
+.. c:function:: int npy_clear_floatstatus(char*)
```
Comment (eric-wieser) [inline doc/source/reference/c-api.coremath.rst]: Should this be `npy_clear_floatstatus_barrier`?

File: numpy/core/src/npymath/ieee754.c.src [hunk 4]
```diff
@@ -574,18 +584,24 @@ npy_longdouble npy_nextafterl(npy_longdouble x, npy_longdouble y)
...
 #include <ieeefp.h>
 
-int npy_get_floatstatus(void)
+int npy_get_floatstatus_barrier(char * param))
 {
     int fpstatus = fpgetsticky();
+    /*
+     * By using a volatile, the compiler cannot reorder this call
+     */
+    if (param != NULL) {
```
Comment (eric-wieser) [inline numpy/core/src/npymath/ieee754.c.src]: Why do this check, yet pass `x` in from `npy_get_floatstatus`?

Comment (mattip) [reply numpy/core/src/npymath/ieee754.c.src]: The correct thing to do is call `npy_get_floatstatus_barrier` directly with a local variable, this currently prevents reordering the call. See for instance `_check_ufunc_fperr` in `numpy/core/src/umath/extobj.c` (where extobj may be NULL), or line 866 in `scalarmath.c.src` which was the original place the reordering was noticed. 

When I fix the documentation from the comment [above](https://github.com/numpy/numpy/pull/11036#pullrequestreview-118975895) I will expand why the `_barrier` form is preferable.

Comment (charris) [reply numpy/core/src/npymath/ieee754.c.src]: I think @eric-wieser was asking why the check was needed at all.

**Comment by charris** [comment]
@mattip This is a horror to backport because it touches stuff all over the place. Could you give it a shot? Leave out the documentation stuff and just get the fix and tests backported.

**Comment by charris** [comment]
I'll give a backport another shot also.

**Comment by charris** [comment]
OK, I squashed the commits and did a backport, looks OK so far.

**Pull request merged**